### ISSUE TITLE
`@alterior/annotations` `EntityAnnotation.decorator` is not type-safe on valid targets

### ISSUE THREAD

**Issue by Woodz** [alterior-mvc/alterior#102]
`EntityAnnotation.decorator` [currently supports limiting the valid targets](https://github.com/alterior-mvc/alterior/tree/main/packages/annotations#limiting-the-decorators-allowed-targets) (e.g. `class`, etc) but this is only enforced at runtime.

However, Typescript decorators are typed to be specific to the allowed targets, meaning that mistakes are caught at "compile" time. This provides a much cleaner DX because errors are caught earlier.

Therefore I propose to narrow the return type of `EntityAnnotation.decorator` based on `validTargets` such that the following mistake is caught at "compile" time, not at runtime.

```
import { Annotation, MetadataName } from '@alterior/annotations';

@MetadataName('foo')
class MyAnnotation extends Annotation {
  constructor(readonly name: string) {
    super();
  }
}

const MyClassDecorator = EntityAnnotation.decorator({validTargets: ['class']});

class TestClass {
  @MyClassDecorator('bar')
  get foo(): number {
    return 1;
  }
}
```

**Comment by rezonant** [comment]
Ooh! Thanks for using Alterior! Sorry I missed this until now. 

Great suggestion! I do believe this should be possible with some clever overloading. 

I think since this is a correctness improvement it's safe to do this during the v3 lifecycle.

**Comment by rezonant** [comment]
This is now available in `@alterior/annotations@3.11.0`. Thanks for the suggestion!
